### PR TITLE
ENH: Support multiple opening hours intervals for BusinessHour

### LINKED ISSUE #15481: ENH: BusinessHour to support multiple opening and closing times per day
BusinessHour supports arbitrary start and end times.  But some businesses open and close more than once per day.  Some examples:

 * Restaurants, who might want to analyze foot traffic or revenue, but excluding times they don't open.
 * Stock markets, several of which have a lunch break.
 * Mexico, where standard business hours are 07:00-14:00 and 16:00-18:00.

Many use cases could be satisfied by just two opens and two closes per day, but it might be just as well to support any amount.

### PR DESCRIPTION
**Pull request by howsiwei**
- [x] closes #15481
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 1]
```diff
@@ -769,6 +775,14 @@ def test_constructor_errors(self):
...
         with pytest.raises(ValueError):
             BusinessHour(start='14:00:05')
+        with pytest.raises(ValueError):
+            BusinessHour(start=[])
+        with pytest.raises(ValueError):
+            BusinessHour(start=['09:00', '11:00'])
+        with pytest.raises(ValueError):
+            BusinessHour(start=['09:00', '11:00'], end=['10:00'])
+        with pytest.raises(ValueError):
+            BusinessHour(start=['09:00', '11:00'], end=['12:00', '20:00'])
```
Comment (gfyoung) [inline pandas/tests/tseries/offsets/test_offsets.py]: Can you update the `pytest.raises` calls (for this entire test function) to include a check for the error message?  Just pass in `match` parameter that indicates the expected message.

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -603,61 +634,66 @@ def next_bday(self):
...
+                                earliest_start.hour, earliest_start.minute)
+            elif self.n * dir < 0 and other.time() < earliest_start:
+                other = other + dir * self.next_bday
+                return datetime(other.year, other.month, other.day,
+                                latest_start.hour, latest_start.minute)
+        if self.n * dir >= 0:
+            for st in self.start:
+                if other.time() <= st:
+                    return datetime(other.year, other.month, other.day,
+                                    st.hour, st.minute)
```
Comment (gfyoung) [inline pandas/tseries/offsets.py]: Comment your code to explain what all of this logic is for.

File: pandas/tests/tseries/offsets/test_offsets.py (hunk 1 above)
Comment (howsiwei) [reply pandas/tests/tseries/offsets/test_offsets.py]: Updated

File: pandas/tseries/offsets.py (hunk 2 above)
Comment (howsiwei) [reply pandas/tseries/offsets.py]: Comments added

File: pandas/tseries/offsets.py [hunk 3]
```diff
@@ -579,9 +580,43 @@ class BusinessHourMixin(BusinessMixin):
 
     def __init__(self, start='09:00', end='17:00', offset=timedelta(0)):
         # must be validated here to equality check
-        start = liboffsets._validate_business_time(start)
+        if _iterable_not_string(start):
+            start = np.asarray(start)
+            if len(start) == 0:
+                raise ValueError('number of starting time cannot be 0')
```
Comment (mroeschke) [inline pandas/tseries/offsets.py]: `"Must include at least 1 start time"`

File: pandas/tseries/offsets.py [hunk 4]
```diff
@@ -579,9 +580,43 @@ class BusinessHourMixin(BusinessMixin):
...
+        else:
+            start = np.array([start])
+        if _iterable_not_string(end):
+            end = np.asarray(end)
+            if len(end) == 0:
+                raise ValueError('number of ending time cannot be 0')
```
Comment (mroeschke) [inline pandas/tseries/offsets.py]: `"Must include at least 1 end time"`

File: pandas/tseries/offsets.py [hunk 5]
```diff
@@ -579,9 +580,43 @@ class BusinessHourMixin(BusinessMixin):
...
+        start = tuple(start[index])
+        end = tuple(end[index])
+
+        total_secs = 0
+        for i in range(num_openings):
+            total_secs += self._get_business_hours_by_sec(start[i], end[i])
+            total_secs += self._get_business_hours_by_sec(
+                end[i], start[(i + 1) % num_openings])
+        if total_secs != 24 * 60 * 60:
+            raise ValueError('invalid starting and ending time(s)')
```
Comment (mroeschke) [inline pandas/tseries/offsets.py]: Can you make this error message a little more informative?

File: pandas/tseries/offsets.py [hunk 6]
```diff
@@ -603,61 +638,74 @@ def next_bday(self):
...
-            raise ValueError('start and end must not be the same')
-        elif self.start < self.end:
-            return True
-        else:
-            return False
+    def _get_daytime_flag(self, start, end):
+        return start < end
 
-    def _next_opening_time(self, other):
+    def _next_opening_time(self, other, dir=1):
```
Comment (mroeschke) [inline pandas/tseries/offsets.py]: Could you use a different variable besides `dir`? (It's a python keyword)

File: pandas/tseries/offsets.py [hunk 7]
```diff
@@ -687,11 +733,15 @@ def rollforward(self, dt):
                 return self._prev_opening_time(dt)
         return dt
 
+    def _get_closing_time(self, dt):
+        for i, st in enumerate(self.start):
+            if st.hour == dt.hour and st.minute == dt.minute:
+                return dt + timedelta(
+                    seconds=self._get_business_hours_by_sec(st, self.end[i]))
+        raise ValueError("dt should be a starting time")
```
Comment (mroeschke) [inline pandas/tseries/offsets.py]: Can you add a test where this is hit?

Comment (howsiwei) [reply pandas/tseries/offsets.py]: You mean where `ValueError` is raised here? `ValueError` would never be raised here if only public methods are called.

File: pandas/tseries/offsets.py (hunk 4 above)
Comment (howsiwei) [reply pandas/tseries/offsets.py]: Updated

File: pandas/tseries/offsets.py (hunk 5 above)
Comment (howsiwei) [reply pandas/tseries/offsets.py]: Updated

File: pandas/tseries/offsets.py (hunk 6 above)
Comment (howsiwei) [reply pandas/tseries/offsets.py]: Updated

File: pandas/tseries/offsets.py (hunk 3 above)
Comment (howsiwei) [reply pandas/tseries/offsets.py]: Updated

File: pandas/tseries/offsets.py (hunk 7 above)
Comment (mroeschke) [reply pandas/tseries/offsets.py]: Well can this `ValueError` get floated up to the user? If so, we should have a test to catch it.

Comment (howsiwei) [reply pandas/tseries/offsets.py]: Only if user calls `_get_closing_time` him/herself

Comment (howsiwei) [reply pandas/tseries/offsets.py]: @mroeschke so should this be tested? If yes, do we just run `_get_closing_time` directly in the test function?

Comment (mroeschke) [reply pandas/tseries/offsets.py]: So if this error can never be hit by the user using public methods, I'm still not sure if we entirely need this error.

Comment (howsiwei) [reply pandas/tseries/offsets.py]: Should I remove the raise statement? Or add a comment to indicate that it would never be hit?

Comment (mroeschke) [reply pandas/tseries/offsets.py]: If `if st.hour == dt.hour and st.minute == dt.minute:` is always guaranteed to get hit, then you can remove the error.

Comment (howsiwei) [reply pandas/tseries/offsets.py]: I have removed the error and added a comment.

**Comment by howsiwei** [comment]
For some reason this PR is automatically closed after rebasing on latest master. I have created a new PR at #26628 and have edited as per all the comments here.

**Pull request closed without merging**